### ISSUE TITLE
Worker crashes and model is reloaded when a gRPC client aborts a prediction request

### ISSUE THREAD

**Issue by bergerkiller** [pytorch/serve#1888]
### 🐛 Describe the bug

We are hosting a GPU model using pytorch-serve, and have gRPC client making requests to this server to perform inference on images. [This uses this proto specification](https://github.com/pytorch/serve/blob/master/frontend/server/src/main/resources/proto/inference.proto).

In production we discovered a DOS situation where the server was endlessly reloading the model after a gRPC client aborts/times out a request to the server. As the model takes a while to be loaded into the GPU, this caused requests to back up even more, which in turn caused more clients to time out as the server couldn't meet throughput.

End result was a server endlessly receiving client disconnects, reloading the model, and all systems making requests receiving timeout errors.

### Error logs

```
2022-10-05T11:40:50,349 [WARN ] W-9000-car_test_model_22.10.276 org.pytorch.serve.wlm.WorkerThread - Backend worker thread exception.
io.grpc.StatusRuntimeException: CANCELLED: call already cancelled
        at io.grpc.Status.asRuntimeException(Status.java:524) ~[model-server.jar:?]
        at io.grpc.stub.ServerCalls$ServerCallStreamObserverImpl.onNext(ServerCalls.java:335) ~[model-server.jar:?]
        at org.pytorch.serve.job.GRPCJob.response(GRPCJob.java:66) ~[model-server.jar:?]
        at org.pytorch.serve.wlm.BatchAggregator.sendResponse(BatchAggregator.java:74) ~[model-server.jar:?]
        at org.pytorch.serve.wlm.WorkerThread.run(WorkerThread.java:195) [model-server.jar:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]
        at java.lang.Thread.run(Thread.java:829) [?:?]
2022-10-05T11:40:50,349 [INFO ] W-9000-car_test_model_22.10.276 ACCESS_LOG - /10.0.0.123:49350 "gRPC org.pytorch.serve.grpc.inference.InferenceAPIsService/Predictions HTTP/2.0" 13 317
2022-10-05T11:40:50,350 [INFO ] W-9000-car_test_model_22.10.276-stdout MODEL_LOG - Frontend disconnected.
2022-10-05T11:40:50,350 [DEBUG] W-9000-car_test_model_22.10.276 org.pytorch.serve.wlm.WorkerThread - W-9000-car_test_model_22.10.276 State change WORKER_MODEL_LOADED -> WORKER_STOPPED
```

### Installation instructions

Using docker installation with .mar file added to model-store using a build script. This is a custom docker installation as it also installs some required third-party libraries (MMCV)

### Model Packaing

Fill a zip file with the metadata file, config, checkpoint, and python handler file. Custom packaging setup part of a CI pipeline.

Python handler is used.

### config.properties

```
inference_address=http://0.0.0.0:8080
management_address=http://0.0.0.0:8081
metrics_address=http://0.0.0.0:8082
number_of_netty_threads=32
job_queue_size=1000
model_store=/home/model-server/model-store
load_models=all
```

### Versions

Had some trouble running this script in our environment, but these are the relevant versions from pip:
```
torch                  1.6.0
torch-model-archiver   0.6.0
torchserve             0.6.0
torchvision            0.7.0
```

And this is the output that did work:
```
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration:
GPU 0: Quadro P1000
Nvidia driver version: 440.82
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
```

### Repro instructions

1. Run pytorch serve with a model which takes ~5s to load up into the gpu
2. Run a python gRPC inference client to talk with this server to perform inference
3. Run the client for roughly 500ms-1s so it gets time to make the request to the server
4. Ctrl+C/Kill the client so it cancels the request
5. Observe the server printing the error
6. Make a request again
7. Observe that the request still takes 5+ seconds as the model is reloaded on the server

You can repeat 1-4 to create a denial of service attack, intentional or not.

### Possible Solution

Don't kill the worker process when a gRPC call is cancelled/killed.

**Comment by msaroufim** [comment]
Which torchserve version are you using? This was fixed here https://github.com/pytorch/serve/issues/1815 and you can try out our nightlies to confirm. Official release coming in November

**Comment by bergerkiller** [comment]
> This was fixed here #1815 and you can try out our nightlies to confirm or build Dockerfile.dev. Official release coming in November

Thanks, I will give this a try tomorrow

**Comment by bergerkiller** [comment]
It was a lot of trouble to get the dev build working because of incompatibilities with MMCV, so I had to tweak things like using the devel cudnn base tag among other things. But, I got it all working, and I can no longer crash the worker process by killing the client. So it looks like the issue is indeed resolved.

Thanks for the quick advice :)
